hudi-common/src/main/java/org/apache/hudi/common/model/HoodieMerge.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.common.model;
+
+import org.apache.avro.Schema;
+import org.apache.hudi.common.util.Option;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Properties;
+
+/**
+ * HoodieMerge defines how to merge two records. It is a stateless component.
+ * It can implement the merging logic of HoodieRecord of different engines
+ * and avoid the performance consumption caused by the serialization/deserialization of Avro payload.
+ */
+public interface HoodieMerge extends Serializable {
+  
+  HoodieRecord preCombine(HoodieRecord older, HoodieRecord newer);
```

Comment:
Thank you for your suggestion. It's very constructive. let`s mull this over more

Reply:
So, historically, this has been 2 different methods with (potentially) different semantics:
  - `preCombine` is de-duplicating the input batch (before inserting it into the table
  - `combineAndGet` is used to merge persisted version with the incoming (that could have been previously de-duplicated

I also don't see a reason for us to get hung up on this historical context and we should try to unify these historically (potentially) divergent methods into 1 providing a single avenue of merging records either inside a batch (when de-duping) or when combining persisted one with the incoming.

Reply:
Thank you for your suggestion. I create a jira to track this. https://issues.apache.org/jira/browse/HUDI-4347

Reply:
Including the JIAR mentioned by @wzx140 , We added two JIRA to track the content to be improved.

Reply:
Including the jiar mentioned by @wzx140  , We added two jiras to track the content to be improved:                              
1. [Detect HoodieMerge type with engine type automatically instead of the default avro-based one ](https://issues.apache.org/jira/browse/HUDI-4301) 
2. [Unify one method in HoodieMerge to replace "preCombine" and "combineAndGetUpdateValue" ](https://issues.apache.org/jira/browse/HUDI-4347)

About 1, we will complete it in Step3; 
About 2, it may be completed after rfc46; 
So as not to block the performance test and subsequent plan of rfc46，So that we can merge Step2 and continue to adjust in Step3.
@xushiyan @vinothchandar @alexeykudinkin  What do you think?

Reply:
@wulei0302 i think we need to make sure that we wrap up RFC-46 with the following artifacts on hands:

 - New APIs (forward-looking, one unified merging API)
 - Legacy APIs (backward-compatible, with `preCombine` and `combineAndGetUpdateValue`, necessary to facilitate the migration onto the new API)

That way, we can encourage folks to migrate from existing model where Legacy API will be providing compatibility out-of-the box onto a new API which we can call a standard and deprecate a legacy one in 0.13